src/frontend/src/handler/create_index.rs

```diff
@@ -77,77 +76,114 @@ pub(crate) fn gen_create_index_plan(
+            }
+            _ => {
+                return Err(ErrorCode::NotSupported(
+                    "index columns should be columns or functions".into(),
+                    "use columns or functions instead".into(),
+                )
+                .into())
+            }
+        };
+        index_columns_ordered_expr.push((expr_impl, order_type));
+    }
 
-    let mut include_columns = if include.is_empty() {
+    if include.is_empty() {
         // Create index to include all (non-hidden) columns by default.
-        table
+        include_columns_expr = table
             .columns()
             .iter()
             .enumerate()
             .filter(|(_, column)| !column.is_hidden)
-            .map(|(x, _)| x)
-            .collect_vec()
+            .map(|(x, column)| {
+                ExprImpl::InputRef(InputRef::new(x, column.column_desc.data_type.clone()).into())
+            })
+            .collect_vec();
     } else {
-        include
-            .iter()
-            .map(to_column_indices)
-            .try_collect::<_, Vec<_>, RwError>()?
+        for column in include {
+            let expr_impl =
+                binder.bind_expr(risingwave_sqlparser::ast::Expr::Identifier(column))?;
+            include_columns_expr.push(expr_impl);
+        }
     };
 
-    let distributed_by_columns = distributed_by
-        .iter()
-        .map(to_column_indices)
-        .try_collect::<_, Vec<_>, RwError>()?;
+    for column in distributed_by {
+        let expr_impl = binder.bind_expr(risingwave_sqlparser::ast::Expr::Identifier(column))?;
+        distributed_columns_expr.push(expr_impl);
+    }
+
+    let table_desc = Rc::new(table.table_desc());
 
     // Remove duplicate column of index columns
     let mut set = HashSet::new();
-    index_columns = index_columns
+    index_columns_ordered_expr = index_columns_ordered_expr
         .into_iter()
-        .filter(|x| set.insert(x.column_index))
+        .filter(|(expr, _)| match expr {
+            ExprImpl::InputRef(input_ref) => set.insert(input_ref.index),
+            ExprImpl::FunctionCall(_) => true,
+            _ => unreachable!(),
+        })
         .collect_vec();
 
     // Remove include columns are already in index columns
-    include_columns = include_columns
+    include_columns_expr = include_columns_expr
         .into_iter()
-        .filter(|x| set.insert(*x))
+        .filter(|expr| match expr {
+            ExprImpl::InputRef(input_ref) => set.insert(input_ref.index),
+            _ => unreachable!(),
+        })
         .collect_vec();
 
     // Remove duplicate columns of distributed by columns
-    let distributed_by_columns = distributed_by_columns.into_iter().unique().collect_vec();
+    let mut set = HashSet::new();
+    let distributed_columns_expr = distributed_columns_expr
+        .into_iter()
+        .filter(|expr| match expr {
+            ExprImpl::InputRef(input_ref) => set.insert(input_ref.index),
+            _ => unreachable!(),
```

Comment:
Is it impossible to use a function column as distributed column?

Reply:
As an improvement, yes. Currently, we ban it at parsing, because I want to focus on the main functionality at first. I will support it in the future if we need it.

Reply:
Take the e2e test case `create index idx on t(j->'k')` as an example, what's the distribution key here?

Reply:
The functional column `j->'k'` by default.